10_rest/n0npax/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,166 @@
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"testing"
+
+	"bou.ke/monkey"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseArgs(t *testing.T) {
+
+	os.Args = []string{"", "-s", "map", "-d", "/dev/null", "--port", "1234"}
+	config, err := parseArgs()
+	assert.NoError(t, err)
+	assert.Equal(t, config.port, 1234)
+	assert.Equal(t, config.path, "/dev/null")
+	assert.Equal(t, config.sType, "map")
+}
+
+func TestMainWithFakeArgs(t *testing.T) {
+	fakeExit := func(int) {
+		panic("foo-arg")
+	}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "--foo"}
+	assert.PanicsWithValuef(t, "foo-arg", main, "foo-arg")
+}
+
+func TestMain(t *testing.T) {
+	fakeRunSrv := func(*gin.Engine, int) error {
+		panic("main")
+	}
+	patch := monkey.Patch(runSrv, fakeRunSrv)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "-d", "../../test-data/data.json"}
+	assert.PanicsWithValuef(t, "main", main, "main")
+}
+
+func TestCreateStorer(t *testing.T) {
+	stores := []string{"map", "sync"}
+	for _, v := range stores {
+		v := v
+		t.Run(v, func(t *testing.T) {
+			s, err := createStorer(v)
+			assert.NoError(t, err)
+			_, ok := s.(puppy.Storer)
+			assert.True(t, ok)
+		})
+	}
+}
+
+func TestCreateNotSupportedStorer(t *testing.T) {
+	stores := []string{"foo", "bar"}
+	for _, v := range stores {
+		v := v
+		t.Run(v, func(t *testing.T) {
+			_, err := createStorer(v)
+			assert.Error(t, err)
+		})
+	}
+}
+
+func TestRunSrvFakePoer(t *testing.T) {
+	err := runSrv(nil, -22)
+	assert.Error(t, err)
+}
+
+func TestRunSrv(t *testing.T) {
+	run := func(*gin.Engine, ...string) error {
+		return nil
+	}
+	patch := monkey.Patch((*gin.Engine).Run, run)
+	defer patch.Unpatch()
+
+	err := runSrv(nil, 8822)
+	assert.NoError(t, err)
+}
+
+func TestExitOnError(t *testing.T) {
+	fakeExit := func(int) {}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	var buf bytes.Buffer
+	out = &buf
+
+	exitOnError(errors.New("test"))
+	actual := buf.String()
+	assert.Equal(t, "test\n", actual)
+}
+
+func TestFeedStore(t *testing.T) {
+	file, err := os.Open("../../test-data/data.json")
+	assert.NoError(t, err)
+	ps, err := readPuppies(file)
+	assert.NoError(t, err)
+	s := store.NewMemStore()
+	err = feedStore(s, ps)
+	assert.NoError(t, err)
+	for i := 0; i < len(ps); i++ {
+		p, err := s.ReadPuppy(i)
+		assert.NoError(t, err)
+		assert.Equal(t, i, p.ID)
+	}
+}
+
+func TestFeedStoreCorruptedData(t *testing.T) {
+	file, err := os.Open("../../test-data/data.json")
+	assert.NoError(t, err)
+	ps, err := readPuppies(file)
+	assert.NoError(t, err)
+	ps[0].Value = -1
+	s := store.NewMemStore()
+	err = feedStore(s, ps)
+	assert.Error(t, err)
+}
+
+func TestFeedStoreNegativeVal(t *testing.T) {
+	file, err := os.Open("../../test-data/data.json")
```

Comment:
inline test data and use [`strings.NewReader()`](https://golang.org/pkg/strings/#NewReader).
That way you don't have to worry about execution context / external files.

Reply:
fixed